File: f/fractal.go
package f

import (
	"image"
	"image/color"
	"runtime"
)

const (
	G = float64(2)
	N = 100
)

type Fractal struct {
	Size    int
	Center  Point
	Scale   float64

	Fmap    Fmap
	Cmap    Cmap

	w       int
	h       int
	Img     *image.NRGBA
	pixels  []pixel
}

func NewFractal(cfg Fractal) Fractal {
	w, h := 3*cfg.Size, 2*cfg.Size
	img := image.NewNRGBA(image.Rect(0, 0, w, h))

	return Fractal{
		Size:     cfg.Size,
		Center:   cfg.Center,
		Scale:    cfg.Scale,
		Fmap:     cfg.Fmap,
		Cmap:     cfg.Cmap,

		w:        w,
		h:        h,
		Img:      img,
		pixels:   initPixels(cfg),
	}
}

func (f Fractal) Render() {
	f.RenderParallel()
}

func (f Fractal) RenderSerial() {
	for _, p := range f.pixels {
		f.render(p)
	}
}

func (f Fractal) RenderParallel() {
	nCPU := runtime.NumCPU()
	c := make(chan int, nCPU)

	for i := 0; i < nCPU; i++ {
		off := i*len(f.pixels) / nCPU
		len := (i + 1)*len(f.pixels) / nCPU

		go f.renderSlice(off, len, c)
	}

	for i := 0; i < nCPU; i++ {
		<-c
	}
}

func (f Fractal) renderSlice(i, len int, c chan int) {
	for ; i < len; i++ {
		f.render(f.pixels[i])
	}

	c <- 1
}

func (f Fractal) render(p pixel) {
	r, g, b, a := f.Cmap(f.Fmap(complex(p.x, p.y)))
	f.Img.SetNRGBA(p.k, p.l, color.NRGBA{r, g, b, a})
}

type Point struct {
	X float64
	Y float64
}

File: viewer/fractal.js
function Fractal(fmap) {
  this.fmap = fmap

  // methods
  this.render = function(view) {
    var img  = new ImageData(view.w, view.h)
    var data = img.data

    view.pixels.forEach(function(p) {
      renderPixel(fmap, data, p, view)
    });

    view.ctx.putImageData(img, 0, 0)
  };

  // this.renderPixels = function() {
  // };
}

function renderPixel(fmap, data, p, view) {
    var s  = view.s
    var cx = view.cx; var cy = view.cy;
    var w  = view.w;  var h  = view.h;
    var k  = p.k;     var l  = p.l

    var x = s * ((3.0/w)*k + cx - 1.5) + cx;
    var y = s * ((2.0/h)*l + cy - 1.0) + cy;

    var z = new Complex(x, y);

    var c = fmap(z); // c = {r: 128, g: 128, b: 128, a: 255};

    var s = 4 * p.i;
    data[s+0] = c.r;
    data[s+1] = c.g;
    data[s+2] = c.b;
    data[s+3] = c.a;
}
